src/core/expression/qgsexpressionfunction.cpp

```diff
@@ -1356,6 +1356,32 @@ static QVariant fcnLength( const QVariantList &values, const QgsExpressionContex
   return QVariant( str.length() );
 }
 
+static QVariant fcnLength3D( const QVariantList &values, const QgsExpressionContext *, QgsExpression *parent, const QgsExpressionNodeFunction * )
+{
+  QgsGeometry geom = QgsExpressionUtils::getGeometry( values.at( 0 ), parent );
+
+  if ( geom.type() != QgsWkbTypes::LineGeometry )
```

Comment:
Curved geometries will also return a line geometry type, so you'll need to handle these somehow (eg by segmentizing and then calculating the distance. Or by working out the math for true 3d distance for curves! 😁)

Reply:
Hi @nyalldawson , thanks for the review and for pointing that out. Could it be a correct approach to always cast to QgsCurve and then use curveToLine() to extract a QgsLineString representation? I don't know the "curve" world well.
Anyway If you have a better idea on how to implement this function, we could disregard this PR.

Reply:
I suggest looking into https://github.com/qgis/QGIS/blob/16493ca562d4961471ec5d7af087e6541b88e46d/src/core/geometry/qgsgeometryutils.cpp#L755 you could either alter it to add z/m values or use the base for the lenght and afterward add surface distance. You can see how it is used in QgsCircularString.

I think that would be adequate.

Reply:
@roya0045 thanks for the good suggestion!

Anyway I think "length3D" should behave like the "length" and "perimeter" (and "area") functions.

It seems they don't use `QgsCircularString::length()` to calculate the length of a CircularString, so the calculated value is equal to the value of `geometry.constGet().curveToLine().length()`.

If fact, with a R=100 circonference/circle, in the field calculator:
- `length(geom_from_wkt('CompoundCurve (CircularString (100 200, 200 100, 100 0, 0 100, 100 200))'))` and
- `perimeter(geom_from_wkt('CurvePolygon (CompoundCurve (CircularString (100 200, 200 100, 100 0, 0 100, 100 200)))'))` 

both return 628.3105558829233 (< 2\*π\*R)

like
```python
circonference=QgsGeometry.fromWkt('CompoundCurve (CircularString (100 200, 200 100, 100 0, 0 100, 100 200))')
circonference.constGet().curveToLine().length()
```
or
```python
circonference.length()
```

while
```python
circonference.constGet().length()
```
returns 628.3185307179587 (=2\*π\*r).

Reply:
@nyalldawson, now length3D function handles curved geometries too.

Reply:
Nice work! Though I'm unsure what make the same tests keep failling in Travis and if something is now broken (unrelated to this PR)